LedgerHQ/ledger-live-desktop
# LIVE-590 POC Testing autodisconnect mechanism

> **This can't be merged. It needs first the ledgerjs merge, then the live-common changes and bump, then this pr updated, but this is the only way to test it.**

## 🦒 Context (issues, jira)
https://ledgerhq.atlassian.net/browse/LIVE-590

For context, this is a continuation of the initial task that attempted to release the hold from the singleton transport after a given amount of time. The problem that we initially had with that approach, problem that was identified by our wonderful QA @alalmi-ledger , was that disconnecting after a hard timeout that only relied on apdu exchanges to reset the timeout didn't account for the fact that we have some long running tasks inside `withDevice` calls.

`withDevice` will open a connection to a device and will keep that connection open unless unmounted or specifically disconnected. In certain flows such as adding accounts or sending transactions we would reach the timeout before completing the flow, triggering a disconnect and breaking said flow.

In https://github.com/LedgerHQ/ledgerjs/pull/797 we introduced the concept of freezing the auto-disconnect temporarily while are still inside a `withDevice` call. This pr leverages that functionality, while relying on a commit for live-common that actually uses that new exposed static method [setAllowAutoDisconnect](https://github.com/LedgerHQ/ledger-live-common/commit/b09ed14715d79c52dd440ef900168e415c10e951) to enable/disable the feature.


## How to test this
Same as the original PR, the context of this is still making sure that we don't keep control of the HID connection after we are done with the device. https://github.com/LedgerHQ/ledger-live-desktop/pull/4647 Just be sure that the failing tests (long add accounts on Cosmos, or a complex send) also work.

The device should be connectable after 5 seconds when we are no longer interacting with it from https://repl.ledger.tools/ on chrome.

In particular, check what happens with flows if we leave the device idle for more than 5 seconds after the action is completed. I noticed the add accounts flow will throw an error when we disconnect even though we are no longer interacting with the device. Perhaps other flows have this same behavior and we'd need to adapt the UX to it (ie, if the action is done, ignore the error basically)

## 🖤  Expectations to reach

PR must pass CI, rebase develop if conflicts. Thanks!

- **on QA**: at least one of these two checkboxes must be checked:
- **on delivery**: at least one of these two checkboxes must be checked:
